README


This archive contains the distribution Email-Abstract,
version 3.007:

  unified interface to mail representations

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.008.



Changes
Revision history for Email-Abstract.

3.007     2013-12-31 10:39:14 America/New_York
          fix skip count when MIME::Entity is not present

3.006     2013-09-01 08:41:48 America/New_York
          require Email::Simple 1.998 to allow ->new(\$string)

3.005     2013-07-30 09:57:36 America/New_York
          repackage, update bug tracker, etc.

          provide raw body on multipart Email::MIME messages (thanks,
          bailey@upenn)
          [rt.cpan.org #77624]

3.004     2011-02-18

          If present, MIME::Entity must be v5.501; v5.500 had a regression (or
          a bug fix, depending how you look at it) that broke header-reading.
          While technically older versions that are not 5.500 would work, it is
          much simpler to just require the newest version, rather than to
          support a version range with a hole in it.

3.003     2010-11-08
          replace Class::ISA with mro (MRO::Compat)

3.002     2010-06-11
          avoid a warning in MailInternet with zero headers found

3.001     2008-12-11
          declare our prereq on Scalar::Util

3.000     2008-12-09
          accept scalar references for more efficient msg-from-string

2.134     2007-11-16
          (no code changes from previous dev release)

2.133_05  2007-11-11
          [BUG FIXES]
          added is_available method to MIMEEntity plugin

2.133_04  2007-09-24
          [ENHANCEMENTS]
          created Email::Abstract::Plugin base class; please use it!
          added is_available method to plugins

          [BUG FIXES]
          is_available in the Mail::Internet adapter should solve header
            folding issues (by preventing you from using it when it can't work)

2.133_03  2007-08-??
          diagnostics in output to indicate what version of a module we used

2.133_02  2007-07-??
          fix test planning

2.133_01  2007-07-??
          add test to ensure that "can't handle" exception is thrown ASAP
          remove unexplained requirement for perl 5.6
          fix Mail::Internet header fetching to unfold headers
          fix Mail::Message body setter, which hosed newlines
          fix body handling for Mail::Internet
          improved consistency of method used to find adapter class
          improved tests and test coverage

2.132     2007-03-22
          packaging improvements

2.131     2006-08-22
          pod tests

2.13      2006-07-24
          test for and permit passing Email::Abstract objects to Email::Abstract
          class methods

2.12      2006-07-24
          don't use MIME::Entity in test if it's not available

2.11      2006-07-22
          better test planning

2.10      2006-07-21
          add a new method to create wrapper objects
          handle subclasses /properly/ (correct ISA order)
          improved tests and test coverage
          miscellaneous refactoring
          update PEP URL
          update documentation

2.01      2004-11-04
          Minor Documentation Fix
          Author Change
          PEP Contact Added

2.0       2004-08-25 12:12:37 BST
          Handle subclasses

0.01      2004-05-26 16:47 20
          original version; created by h2xs 1.22 with options
            -AX -b 5.6.0 -n Email::Abstract



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.008.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "unified interface to mail representations",
  "AUTHOR" => "Ricardo SIGNES <rjbs\@cpan.org>, Simon Cozens <simon\@cpan.org>, Casey West <casey\@geeknest.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Email-Abstract",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Email::Abstract",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Email::Simple" => "1.998",
    "MRO::Compat" => 0,
    "Module::Pluggable" => "1.5",
    "Scalar::Util" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.96",
    "lib" => 0
  },
  "VERSION" => "3.007",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Email::Simple" => "1.998",
  "MRO::Compat" => 0,
  "Module::Pluggable" => "1.5",
  "Scalar::Util" => 0,
  "Test::More" => "0.96",
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Email/Abstract/Plugin.pm
use strict;
use warnings;
package Email::Abstract::Plugin;
{
  $Email::Abstract::Plugin::VERSION = '3.006';
}
# ABSTRACT: a base class for Email::Abstract plugins


sub is_available { 1 }

1;

__END__

=pod

=head1 NAME

Email::Abstract::Plugin - a base class for Email::Abstract plugins

=head1 VERSION

version 3.006

=head1 METHODS

=head2 is_available

This method returns true if the plugin should be considered available for
registration.  Plugins that return false from this method will not be
registered when Email::Abstract is loaded.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Abstract/EmailMIME.pm
use strict;
use warnings;
package Email::Abstract::EmailMIME;
{
  $Email::Abstract::EmailMIME::VERSION = '3.006';
}
# ABSTRACT: Email::Abstract wrapper for Email::MIME

use Email::Abstract::EmailSimple;
BEGIN { @Email::Abstract::EmailMIME::ISA = 'Email::Abstract::EmailSimple' };

sub target { "Email::MIME" }

sub construct {
    require Email::MIME;
    my ($class, $rfc822) = @_;
    Email::MIME->new($rfc822);
}

1;

__END__

=pod

=head1 NAME

Email::Abstract::EmailMIME - Email::Abstract wrapper for Email::MIME

=head1 VERSION

version 3.006

=head1 DESCRIPTION

This module wraps the Email::MIME mail handling library with an
abstract interface, to be used with L<Email::Abstract>

=head1 SEE ALSO

L<Email::Abstract>, L<Email::MIME>.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Abstract/MIMEEntity.pm
use strict;
package Email::Abstract::MIMEEntity;
{
  $Email::Abstract::MIMEEntity::VERSION = '3.006';
}
# ABSTRACT: Email::Abstract wrapper for MIME::Entity

use Email::Abstract::Plugin;
BEGIN { @Email::Abstract::MIMEEntity::ISA = 'Email::Abstract::MailInternet' };

my $is_avail;
sub is_available {
  return $is_avail if defined $is_avail;
  eval { require MIME::Entity; MIME::Entity->VERSION(5.501); 1 };
  return $is_avail = $@ ? 0 : 1;
}

sub target { "MIME::Entity" }

sub construct {
    require MIME::Parser;
    my $parser = MIME::Parser->new;
    $parser->output_to_core(1);
    my ($class, $rfc822) = @_;
    $parser->parse_data($rfc822);
}

sub get_body {
  my ($self, $obj) = @_;
  my $handle = $obj->bodyhandle;
  return $handle ? $handle->as_string : join('', @{ $obj->body });
}

sub set_body {
    my ($class, $obj, $body) = @_;
    my @lines = split /\n/, $body;
    my $io = $obj->bodyhandle->open("w");
    foreach (@lines) { $io->print($_."\n") }
    $io->close;
}

1;

__END__

=pod

=head1 NAME

Email::Abstract::MIMEEntity - Email::Abstract wrapper for MIME::Entity

=head1 VERSION

version 3.006

=head1 DESCRIPTION

This module wraps the MIME::Entity mail handling library with an
abstract interface, to be used with L<Email::Abstract>

=head1 SEE ALSO

L<Email::Abstract>, L<MIME::Entity>.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Abstract/EmailSimple.pm
use strict;
use warnings;
package Email::Abstract::EmailSimple;
{
  $Email::Abstract::EmailSimple::VERSION = '3.006';
}
# ABSTRACT: Email::Abstract wrapper for Email::Simple

use Email::Abstract::Plugin;
BEGIN { @Email::Abstract::EmailSimple::ISA = 'Email::Abstract::Plugin' };

sub target { "Email::Simple" }

sub construct {
    require Email::Simple;
    my ($class, $rfc822) = @_;
    Email::Simple->new($rfc822);
}

sub get_header {
    my ($class, $obj, $header) = @_;
    $obj->header($header);
}

sub get_body {
    my ($class, $obj) = @_;
    $obj->body();
}

sub set_header {
    my ($class, $obj, $header, @data) = @_;
    $obj->header_set($header, @data);
}

sub set_body   {
    my ($class, $obj, $body) = @_;
    $obj->body_set($body);
}

sub as_string {
    my ($class, $obj) = @_;
    $obj->as_string();
}

1;

__END__

=pod

=head1 NAME

Email::Abstract::EmailSimple - Email::Abstract wrapper for Email::Simple

=head1 VERSION

version 3.006

=head1 DESCRIPTION

This module wraps the Email::Simple mail handling library with an
abstract interface, to be used with L<Email::Abstract>

=head1 SEE ALSO

L<Email::Abstract>, L<Email::Simple>.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Abstract/MailMessage.pm
use strict;
package Email::Abstract::MailMessage;
{
  $Email::Abstract::MailMessage::VERSION = '3.006';
}
# ABSTRACT: Email::Abstract wrapper for Mail::Message

use Email::Abstract::Plugin;
BEGIN { @Email::Abstract::MailMessage::ISA = 'Email::Abstract::Plugin' };

sub target { "Mail::Message" }

sub construct {
    require Mail::Message;
    my ($class, $rfc822) = @_;
    Mail::Message->read($rfc822);
}

sub get_header {
    my ($class, $obj, $header) = @_;
    $obj->head->get($header);
}

sub get_body {
    my ($class, $obj) = @_;
    $obj->decoded->string;
}

sub set_header {
    my ($class, $obj, $header, @data) = @_;
    $obj->head->delete($header);
    $obj->head->add($header, $_) for @data;
}

sub set_body {
    my ($class, $obj, $body) = @_;
    $obj->body(Mail::Message::Body->new(data => $body));
}

sub as_string {
    my ($class, $obj) = @_;
    $obj->string;
}

1;

__END__

=pod

=head1 NAME

Email::Abstract::MailMessage - Email::Abstract wrapper for Mail::Message

=head1 VERSION

version 3.006

=head1 DESCRIPTION

This module wraps the Mail::Message mail handling library with an
abstract interface, to be used with L<Email::Abstract>

=head1 SEE ALSO

L<Email::Abstract>, L<Mail::Message>.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Abstract/MailInternet.pm
use strict;
package Email::Abstract::MailInternet;
{
  $Email::Abstract::MailInternet::VERSION = '3.006';
}
# ABSTRACT: Email::Abstract wrapper for Mail::Internet

use Email::Abstract::Plugin;
BEGIN { @Email::Abstract::MailInternet::ISA = 'Email::Abstract::Plugin' };

sub target { "Mail::Internet" }

# We need 1.77 because otherwise headers unfold badly.
my $is_avail;
sub is_available {
  return $is_avail if defined $is_avail;
  require Mail::Internet;
  eval { Mail::Internet->VERSION(1.77) };
  return $is_avail = $@ ? 0 : 1;
}

sub construct {
    require Mail::Internet;
    my ($class, $rfc822) = @_;
    Mail::Internet->new([ map { "$_\x0d\x0a" } split /\x0d\x0a/, $rfc822]);
}

sub get_header {
    my ($class, $obj, $header) = @_;
    my @values = $obj->head->get($header);
    return unless @values;

    # No reason to s/// lots of values if we're just going to return one.
    $#values = 0 if not wantarray;

    chomp @values;
    s/(?:\x0d\x0a|\x0a\x0d|\x0a|\x0d)\s+/ /g for @values;

    return wantarray ? @values : $values[0];
}

sub get_body {
    my ($class, $obj) = @_;
    join "", @{$obj->body()};
}

sub set_header {
    my ($class, $obj, $header, @data) = @_;
    my $count = 0;
    $obj->head->replace($header, shift @data, ++$count) while @data;
}

sub set_body {
    my ($class, $obj, $body) = @_;
    $obj->body( map { "$_\n" } split /\n/, $body );
}

sub as_string { my ($class, $obj) = @_; $obj->as_string(); }

1;

__END__

=pod

=head1 NAME

Email::Abstract::MailInternet - Email::Abstract wrapper for Mail::Internet

=head1 VERSION

version 3.006

=head1 DESCRIPTION

This module wraps the Mail::Internet mail handling library with an
abstract interface, to be used with L<Email::Abstract>

=head1 SEE ALSO

L<Email::Abstract>, L<Mail::Internet>.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Abstract.pm
use 5.006;
use warnings;
use strict;
package Email::Abstract;
{
  $Email::Abstract::VERSION = '3.006';
}
# ABSTRACT: unified interface to mail representations
use Carp;
use Email::Simple;
use MRO::Compat;

use Module::Pluggable 1.5
  search_path => [__PACKAGE__],
  except      => 'Email::Abstract::Plugin',
  require     => 1;

use Scalar::Util ();

my @plugins = __PACKAGE__->plugins(); # Requires them.
my %adapter_for =
  map  { $_->target => $_ }
  grep {
    my $avail = eval { $_->is_available };
    $@ ? ($@ =~ /Can't locate object method "is_available"/) : $avail;
  }
  @plugins;

sub object {
  my ($self) = @_;
  return unless ref $self;
  return $self->[0];
}

sub new {
  my ($class, $foreign) = @_;

  return $foreign if eval { $foreign->isa($class) };

  $foreign = Email::Simple->new($foreign)
    unless Scalar::Util::blessed($foreign);

  my $adapter = $class->__class_for($foreign); # dies if none available
  return bless [ $foreign, $adapter ] => $class;
}

sub __class_for {
  my ($self, $foreign, $method, $skip_super) = @_;
  $method ||= 'handle';

  my $f_class = ref $foreign;
     $f_class = $foreign unless $f_class;

  return $f_class if ref $foreign and $f_class->isa($self);

  return $adapter_for{$f_class} if $adapter_for{$f_class};

  if (not $skip_super) {
    my @bases = @{ mro::get_linear_isa($f_class) };
    shift @bases;
    for my $base (@bases) {
      return $adapter_for{$base} if $adapter_for{$base};
    }
  }

  Carp::croak "Don't know how to $method $f_class";
}

sub _adapter_obj_and_args {
  my $self = shift;

  if (my $thing = $self->object) {
    return ($self->[1], $thing, @_);
  } else {
    my $thing   = shift;
    my $adapter = $self->__class_for(
      Scalar::Util::blessed($thing) ? $thing : 'Email::Simple'
    );
    return ($adapter, $thing, @_);
  }
}

for my $func (qw(get_header get_body set_header set_body as_string)) {
  no strict 'refs';
  *$func = sub {
    my $self = shift;
    my ($adapter, $thing, @args) = $self->_adapter_obj_and_args(@_);

    # In the event of Email::Abstract->get_body($email_abstract), convert
    # it into an object method call.
    $thing = $thing->object if eval { $thing->isa($self) };

    # I suppose we could work around this by leaving @_ intact and assigning to
    # it.  That seems ... not good. -- rjbs, 2007-07-18
    unless (Scalar::Util::blessed($thing)) {
      Carp::croak "can't alter string in place" if substr($func, 0, 3) eq 'set';
      $thing = Email::Simple->new(
        ref $thing ? \do{my$str=$$thing} : $thing
      );
    }

    return $adapter->$func($thing, @args);
  };
}

sub cast {
  my $self = shift;
  my ($from_adapter, $from, $to) = $self->_adapter_obj_and_args(@_);

  my $adapter = $self->__class_for($to, 'construct', 1);

  my $from_string = ref($from) ? $from_adapter->as_string($from) : $from;

  return $adapter->construct($from_string);
}

1;

__END__

=pod

=head1 NAME

Email::Abstract - unified interface to mail representations

=head1 VERSION

version 3.006

=head1 SYNOPSIS

  my $message = Mail::Message->read($rfc822)
             || Email::Simple->new($rfc822)
             || Mail::Internet->new([split /\n/, $rfc822])
             || ...
             || $rfc822;

  my $email = Email::Abstract->new($message);

  my $subject = $email->get_header("Subject");
  $email->set_header(Subject => "My new subject");

  my $body = $email->get_body;

  $rfc822 = $email->as_string;

  my $mail_message = $email->cast("Mail::Message");

=head1 DESCRIPTION

C<Email::Abstract> provides module writers with the ability to write
simple, representation-independent mail handling code. For instance, in the
cases of C<Mail::Thread> or C<Mail::ListDetector>, a key part of the code
involves reading the headers from a mail object. Where previously one would
either have to specify the mail class required, or to build a new object from
scratch, C<Email::Abstract> can be used to perform certain simple operations on
an object regardless of its underlying representation.

C<Email::Abstract> currently supports C<Mail::Internet>, C<MIME::Entity>,
C<Mail::Message>, C<Email::Simple>, C<Email::MIME>, and C<Courriel>.  Other
representations are encouraged to create their own C<Email::Abstract::*> class
by copying C<Email::Abstract::EmailSimple>.  All modules installed under the
C<Email::Abstract> hierarchy will be automatically picked up and used.

=head1 METHODS

All of these methods may be called either as object methods or as class
methods.  When called as class methods, the email object (of any class
supported by Email::Abstract) must be prepended to the list of arguments, like
so:

  my $return = Email::Abstract->method($message, @args);

This is provided primarily for backwards compatibility.

=head2 new

  my $email = Email::Abstract->new($message);

Given a message, either as a string or as an object for which an adapter is
installed, this method will return a Email::Abstract object wrapping the
message.

If the message is given as a string, it will be used to construct an object,
which will then be wrapped.

=head2 get_header

  my $header  = $email->get_header($header_name);

  my @headers = $email->get_header($header_name);

This returns the values for the given header.  In scalar context, it returns
the first value.

=head2 set_header

  $email->set_header($header => @values);

This sets the C<$header> header to the given one or more values.

=head2 get_body

  my $body = $email->get_body;

This returns the body as a string.

=head2 set_body

  $email->set_body($string);

This changes the body of the email to the given string.

B<WARNING!>  You probably don't want to call this method, despite what you may
think.  Email message bodies are complicated, and rely on things like content
type, encoding, and various MIME requirements.  If you call C<set_body> on a
message more complicated than a single-part seven-bit plain-text message, you
are likely to break something.  If you need to do this sort of thing, you
should probably use a specific message class from end to end.

This method is left in place for backwards compatibility.

=head2 as_string

  my $string = $email->as_string;

This returns the whole email as a decoded string.

=head2 cast

  my $mime_entity = $email->cast('MIME::Entity');

This method will convert a message from one message class to another.  It will
throw an exception if no adapter for the target class is known, or if the
adapter does not provide a C<construct> method.

=head2 object

  my $message = $email->object;

This method returns the message object wrapped by Email::Abstract.  If called
as a class method, it returns false.

Note that, because strings are converted to message objects before wrapping,
this method will return an object when the Email::Abstract was constructed from
a string. 

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Simon Cozens <simon@cpan.org>

=item *

Casey West <casey@geeknest.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2004 by Simon Cozens.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
